lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Random.hs

```diff
@@ -77,6 +94,50 @@ import qualified Data.ByteArray as BA
 import qualified Data.ByteString as BS
 import qualified Data.ByteString.Lazy as BL
 
+{-------------------------------------------------------------------------------
+                                   Key Types
+-------------------------------------------------------------------------------}
+
+-- | Data for generating random scheme addresses
+data RndKey (level :: Depth) key = RndKey
+    { masterKey :: Key 'RootK key
+    -- ^ The root key, generated from the seed.
+    , accountKey :: Key 'AccountK key
+    -- ^ The account key, generated from the root key and account index.
+    , addressKey :: Key 'AddressK key
+    -- ^ The address key, generated from the account key and address index.
+    , accountIndex :: Index 'Hardened 'AccountK
+    -- ^ Which account in the hierarchical scheme to generate addresses in.
+    , addressIndex :: Index 'Soft 'AddressK
```

Comment:
A `RndKey` needs to capture a few things:

- The current depth in the hierarchy (Root, Account, Address)
- Whether the key is private or public
- The current derivation path 

So, a possible structure for this would be:

```hs
data RndKey (depth :: Depth) key = RndKey
    { rawKey :: key
    , derivationPath :: DerivationPath depth
    }

type family DerivationPath (depth :: Depth) :: * where
    DerivationPath 'RootK =
        ()
    DerivationPath 'AccountK =
        Index 'Hardened 'AccountK
    DerivationPath 'AddressK =
        (Index 'Hardened 'AccountK, Index 'Hardened 'AccountK)
```

This allows us to represent non-ambiguously the various type of keys we need for the random derivation, without the need for placeholder data.
Now, in order to implement `KeyToAddress`, we need an extra bit of information which is the actual passphrase derived from the public root key associated to a particular address private key. 

I see three options:

1/ Modify `KeyToAddress` to also require an extra argument. However, this need to be an associated type family from the type class since sequential key doesn't require any extra argument. So, something in the lines of:

```hs
class KeyToAddress target key where
    type KeyToAddressArg key :: *
    keyToAddress :: KeyToAddressArg key -> key 'AddressK XPub -> Address

instance KeyToAddress t SeqKey where
     type KeyToAddressArg SeqKey = ()
     ...

instance KeyToAddress t RndKey where
    type KeyToAddressArg RndKey = RndKey 'RootK XPub
    ...
```

2/ Or, we can directly embed the root public key in any `RndKey` we have, and extend the definition with an extra `rootKey` field.

```hs
data RndKey (depth :: Depth) key = RndKey
    { rawKey :: key
    , derivationPath :: DerivationPath depth
    , rootKey :: key -- For derivation payload encryption 
    }
```

or, alternatively (preferred):

3/ Store only the encryption passphrase with the key (i.e., derive it when we generate the key from the seed initially, and keep it there): 

```hs
data RndKey (depth :: Depth) key = RndKey
    { rawKey :: key
    , derivationPath :: DerivationPath depth
    , payloadPassphrase :: Passphrase "address-derivation-payload"
    }
```

Reply:
thanks for this explanation!

Reply:
Thanks, a top-level type synonym family for the derivation path makes it work.